applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp

```diff
@@ -0,0 +1,887 @@
+
+		double phi;
+		if ((dx > numerical_limit) & (std::abs(dy) < numerical_limit)) phi = 0.00; // dy = 0 and dx > 0
+		else if ((dx < -numerical_limit) & (std::abs(dy) < numerical_limit)) phi = Pi; // dy = 0 and dx < 0
+		else if (std::abs(dx) < numerical_limit)
+		{
+			phi = Pi / 2.00; // dy > 0 and dx = 0
+			if (dy < -numerical_limit) phi = 1.500 * Pi;  // dy < 0 and dx = 0
+		}
+		else
+		{
+			phi = (norm - dx) / dy;
+			phi = std::atan(phi);
+			phi = 2.00 * phi;
+		}
+
+		return phi;
+		KRATOS_CATCH("")
+	}
+
+	double CrBeamElement2D2N::CalculateDeformedElementAngle()
+	{
+		KRATOS_TRY;
+		const double numerical_limit = std::numeric_limits<double>::epsilon();
+
+		Vector CurrentDisplacement = ZeroVector(msElementSize);
+		this->GetValuesVector(CurrentDisplacement,0);
+
+		const double dx = (this->GetGeometry()[1].X0()+CurrentDisplacement[3])
+		 - (this->GetGeometry()[0].X0()+CurrentDisplacement[0]);
+		const double dy = (this->GetGeometry()[1].Y0()+CurrentDisplacement[4])
+		 - (this->GetGeometry()[0].Y0()+CurrentDisplacement[1]);
+
+		const double norm = std::sqrt((dx*dx) + (dy*dy));
+
+		double phi;
+		if ((dx > numerical_limit) & (std::abs(dy) < numerical_limit)) phi = 0.00; // dy = 0 and dx > 0
+		else if ((dx < -numerical_limit) & (std::abs(dy) < numerical_limit)) phi = Pi; // dy = 0 and dx < 0
+		else if (std::abs(dx) < numerical_limit)
+		{
+			phi = Pi / 2.00; // dy > 0 and dx = 0
+			if (dy < -numerical_limit) phi = 1.500 * Pi;  // dy < 0 and dx = 0
+		}
+		else
+		{
+			phi = (norm - dx) / dy;
+			phi = std::atan(phi);
+			phi = 2.00 * phi;
+		}
+
+		return phi;
+		KRATOS_CATCH("")
+	}
+
+	double CrBeamElement2D2N::CalculateCurrentLength() 
+	{
+		KRATOS_TRY;
+		const double du = this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_X)
+			- this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_X);
+		const double dv = this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_Y)
+			- this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_Y);
+
+		const double dx = this->GetGeometry()[1].X0() - this->GetGeometry()[0].X0();
+		const double dy = this->GetGeometry()[1].Y0() - this->GetGeometry()[0].Y0();
+
+		const double l = std::sqrt((du + dx)*(du + dx) + (dv + dy)*(dv + dy));
+		return l;
+		KRATOS_CATCH("")
+	}
+
+	double CrBeamElement2D2N::CalculateReferenceLength() {
+		KRATOS_TRY;
+		const double dx = this->GetGeometry()[1].X0() - this->GetGeometry()[0].X0();
+		const double dy = this->GetGeometry()[1].Y0() - this->GetGeometry()[0].Y0();
+		const double L = std::sqrt((dx*dx) + (dy*dy));
+		return L;
+		KRATOS_CATCH("")
+	}
+
+	bounded_matrix<double,CrBeamElement2D2N::msElementSize,
+	CrBeamElement2D2N::msLocalSize> CrBeamElement2D2N::CalculateTransformationS() 
+	{
+		KRATOS_TRY;
+		const double L = this->CalculateCurrentLength();
+		bounded_matrix<double,msElementSize,msLocalSize> S = ZeroMatrix(msElementSize, msLocalSize);
+		S(0, 0) = -1.00;
+		S(1, 2) = 2.00 / L;

```

Comment:
what if `L` is very small? => add check (maybe even in `CalculateCurrentLength`-function)